gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -0,0 +1,455 @@
+    """
+    raise NotImplementedError
+
+  def _get_variant(self, data_line):
+    # type: (str) -> Variant
+    """Converts a single data_line of a VCF file into a Variant object.
+    In case something goes wrong it must return a MalformedVcfRecord object.
+    Note: this method will be called by next(), one line at a time.
+    """
+    raise NotImplementedError
+
+
+class PyVcfParser(VcfParser):
+  """An Iterator for processing a single VCF file using PyVcf."""
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               file_pattern,  # type: str
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               representative_header_lines=None,  # type:  List[str]
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PyVcfParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    self._header_lines = []
+    self._next_line_to_process = None
+    self._current_line = None
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+
+  def _init_with_header(self, header_lines):
+    self._header_lines = header_lines
+    try:
+      self._vcf_reader = vcf.Reader(fsock=self._line_generator())
+    except SyntaxError as e:
+      raise ValueError(
+          'Invalid VCF header in %s: %s' % (self._file_name, str(e)))
+
+  def _get_variant(self, data_line):
+    # _line_generator will consume this line.
+    self._next_line_to_process = data_line
+    try:
+      record = next(self._vcf_reader)
+      return self._convert_to_variant(record, self._vcf_reader.formats)
+    except (LookupError, ValueError) as e:
+      logging.warning('VCF record read failed in %s for line %s: %s',
+                      self._file_name, data_line, str(e))
+      return MalformedVcfRecord(self._file_name, data_line, str(e))
+
+  def _line_generator(self):
+    for header in self._header_lines:
+      yield header
+    # Continue to process the next line indefinitely. The next line is set
+    # inside _get_variant() and this method is indirectly called in get_variant.
+    while self._next_line_to_process:
+      # PyVCF has explicit str() calls when parsing INFO fields, which fails
+      # with UTF-8 decoded strings. Encode the line back to UTF-8.
+      self._current_line = self._next_line_to_process
+      self._next_line_to_process = None
+      yield self._current_line.encode('utf-8')
+      # Making sure _get_variant() assigned a new value before consuming it.
+      assert self._next_line_to_process
```

Comment:
Please add a comment on this assert:
`assert self._next_line_to_process, 'Internal error: A data line is requested to be processed more than once. Please file a bug if you see this!'`

Reply:
Also, I'd relax this assert to `self._next_line_to_process is not None` (I know that we skip empty lines, but even if we didn't, this means that something was processed).

Reply:
done.